### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `T.cumsum`/`T.cummax` over a row-offset 2-D shared sub-region silently scans the wrong rows instead of the requested ones

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2536]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

`0.1.11` (source tag `v0.1.11`, commit `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`).

### System information

NVIDIA H100 80GB HBM3 (sm_90), CUDA 13, PyTorch build from the wheel. The defect is in the architecture-independent region-access-pointer helper (`src/op/utils.cc`), so it is not Hopper-specific; sm_90 is where it was run.

### Problem description

`T.cumsum` (and `T.cummax`) over a 2-D shared sub-region whose outer (row) start is non-zero — e.g. `T.cumsum(tile[64:128, :], dim=1)` — compiles and runs with no error but scans the **wrong rows**: it reads and writes rows `[0:64]` of the buffer while the user asked for rows `[64:128]`. The result is silently wrong and deterministic.

The base pointer passed to the scan device kernel is `&tile[0]` regardless of the region's row offset. The generated call for `tile[64:128, :]` on a `(128, 8)` buffer is:

```
CumSum2D<128, 1, false>::run((&(tile[0])), (&(tile[0])), 64, 8)
```

The extents `64, 8` are correct, but the base pointer should be `&tile[64*W]` (row 64 × pitch 8), not `&tile[0]`.

<details>
<summary>Root cause: the offset loop in <code>MakeAccessPtrFromRegion</code> skips the innermost two dims</summary>

The scan lowering builds its src/dst pointers via [`MakeAccessPtrFromRegion`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/op/utils.cc#L104-L108), which sums the region's per-dim `min * stride` into the base offset — but only over dims `[0 .. ndim-3]`:

```cpp
// src/op/utils.cc:104-108
offset = make_const(buf->shape[0].dtype(), 0);
for (int i = 0; i < ndim - 2; ++i) {          // ndim==2 -> i<0 -> body never runs
  offset = offset + region->region[i]->min * strides[i];
}
```

For a 2-D region (`ndim == 2`) the bound is `i < 0`, so the loop body never runs and `offset` stays `0` — both `region[0]->min` (the row start) and `region[1]->min` (the column start) are discarded. The intent was to leave the innermost two dims to the device kernel's `H, W` handling, but those extents are *sizes*, not *starting coordinates*: the region's `min` on the innermost dims is a base-address concern and must be folded into the pointer regardless.

[`LowerSharedScan`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/backend/common/op/scan.h#L42-L43) is the sole caller (`MakeAccessPtrFromRegion(op.srcRegion_, 1)` / `(op.dstRegion_, 2)`), and it is the shared template for [both `LowerCumSum` and `LowerCumMax`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/backend/common/op/scan.h#L76-L85), so both ops carry the bug.
</details>

### Reproducible example code

```python
import torch, tilelang
import tilelang.language as T

H, W = 128, 8   # inner width == buffer pitch, so no row-pitch effect is in play (see below)

def scan_kernel(r0, r1):
    @T.prim_func
    def kern(A: T.Tensor((H, W), "float32"), B: T.Tensor((H, W), "float32")):
        with T.Kernel(1, threads=128):
            tile = T.alloc_shared((H, W), "float32")
            T.copy(A, tile)
            T.cumsum(tile[r0:r1, :], dim=1)   # scan a row sub-slice: region[0].min = r0
            T.copy(tile, B)
    return kern

def run(r0, r1, tag):
    jit = tilelang.compile(scan_kernel(r0, r1), target="cuda")
    A = torch.arange(H * W, dtype=torch.float32, device="cuda").reshape(H, W) + 1.0
    B = torch.empty_like(A); jit(A, B); torch.cuda.synchronize()
    exp = A.clone(); exp[r0:r1, :] = torch.cumsum(A[r0:r1, :], dim=1)
    print(f"[{tag}] slice=[{r0}:{r1}] match={torch.allclose(B, exp)}")

run(0,  64,  "control r0=0 ")   # region[0].min == 0 -> nothing to drop -> match=True
run(64, 128, "trigger r0=64")   # region[0].min == 64 dropped   -> match=False
```

<details>
<summary>Observed output on H100 (sm_90), plus the wrong-vs-right rows and the discriminating controls</summary>

```
[control r0=0 ] slice=[0:64]   match=True
[trigger r0=64] slice=[64:128]  match=False
   expected row 64 [:8] = [513, 1027, 1542, 2058, 2575, 3093, 3612, 4132]
   got      row 64 [:8] = [513, 514, 515, 516, 517, 518, 519, 520]   # raw input, never scanned
   got      row 0  [:8] = [1, 3, 6, 10, 15, 21, 28, 36]              # got row-64's scan (should be untouched)
```

**Controls that isolate the defect to the dropped row offset:**

- **`r0 == 0` (above):** the identical program with the row start at 0 is bit-exact — the only difference is the non-zero `min`, so the wrong result comes from dropping it.
- **A bare `T.copy(tile[64:128, :], B)` (no scan) of the same region is correct.** The row offset is handled fine on the normal copy lowering; only the scan path — the sole caller of `MakeAccessPtrFromRegion` — drops it. So this is specific to the scan pointer, not a general slicing problem.
- **`T.cummax`** shows the same failure via the same `LowerSharedScan` template — with a non-monotonic input (so `cummax(row) != row`), `tile[64:128, :]` returns the unscanned raw row `64` while row `0` receives the scan.

Deterministic across runs; matrix of shapes/dims/`reverse` all reproduce the same "scanned base row 0 instead of `r`" symptom.
</details>

### Traceback

```pytb
No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic.
```

### Expected behavior

Either resolution is acceptable, but the current silent miscompile is not:

- **Support** the documented region slice: the scan should operate on the requested rows `[r0:r1]` — i.e. the base pointer must include `region[0]->min * stride`. The `r0 == 0` control of the identical program is already correct, and a plain `T.copy` of the same offset region is already correct, so honoring the offset here is the consistent behavior.
- **Reject** at compile time if row-offset scan regions are out of scope, rather than silently producing a wrong result.

### Additional context

**Suggested fix (proposed, not built).** In [`MakeAccessPtrFromRegion`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/op/utils.cc#L104-L108) the offset should sum over all dims, since the `min` of the innermost dims is still a base-address term (the `extent` line already handles the sizes):

```cpp
for (int i = 0; i < ndim; ++i)                 // was: i < ndim - 2
  offset = offset + region->region[i]->min * strides[i];
```

Alternatively, if only zero-offset scan regions are intended, `LowerSharedScan` could `ICHECK` that every `region[i]->min` is zero and reject otherwise.

**Distinct from #2523.** #2523 is the same op family but a different defect: there the base offset is correct and the **row pitch** is wrong (the device kernel uses the slice width as the stride for a width-narrowed, non-contiguous region with `min == 0`). This report is the **base offset** itself (`utils.cc:106`), fires for a row-offset region with `min != 0`, and reproduces with the inner width equal to the buffer pitch (`W == 8 == full pitch`) so the pitch face cannot apply. The two are **orthogonal, independent defects** — different lines (base-offset computation in `utils.cc` vs the device kernel's stride use) and different fixes — but they are not mutually exclusive: a call like `big[64:96, 0:40]` on a `(128, 64)` buffer trips **both** at once (row `min != 0` drops the offset, and the narrowed column width `40 != 64` mis-strides the rows). The `W == pitch` repro here isolates the offset face so it can be verified alone; it does not imply the two cannot co-occur. I searched the open and closed tracker and found no report of this offset face.

**Provenance.** The `i < ndim - 2` offset loop was introduced in [#1321 "[BugFix] Use BufferRegion in tl.cumsum to infer buffer shape"](https://github.com/tile-ai/tilelang/pull/1321) (merged 2025-11-24), originally in `src/op/reduce.cc`; [#1333](https://github.com/tile-ai/tilelang/pull/1333) (merged 2025-11-25) relocated it verbatim to `src/op/utils.cc` (a refactor, not the origin). Traced via `git log -S'i < ndim - 2'` and `gh pr view`.
